### DISCUSSION TITLE
Plans to add `zeroize::Zeroize` traits to in-built `PyType`?

### DISCUSSION THREAD

**Discussion by jymchng** [PyO3/pyo3#3041] [General]
For pedagogical purposes, is there a short introduction on how custom wrapper arounds `PyType`, e.g. `Py<PyString>` are dropped? My intent is to explore the possibility of having the trait `Zeroize` from `zeroize` crate implemented on custom #[pyclass].

**Comment by davidhewitt** [comment]
These are non-null pointers to Python objects, dropping them just decreases the reference count by one. If you don't have the GIL held when drop occurs, PyO3 will remember the reference to decrease later on next GIL acquisition.

If you want to zero-out the pointer I think you'd want `Option<Py<PyString>>`.

**Reply by jymchng** [reply]
Thank you for your reply.

By using `Option<Py<PyString>>`, do you mean if I want to 'zero' out the value of type `PyString` at the end of its usage, e.g. as a parameter to a function call etc, I can simply set the value of the `Option` to its `None` variant?

If it is used as a parameter of type `PyString` to a rust function call, just before the function returns, how would I be able to set it to `Option<Py<PyString>>::None` if it is not of type `Option` in the function signature?

**Reply by davidhewitt** [reply]
Yes, you could zero out by setting to the `None` variant. 

You wouldn't be able to do this when the function signature is not optional. It's not clear to me what use case you're trying to solve by asking for that?
